drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py

```diff
@@ -0,0 +1,590 @@
+                  }
+       [rospy.Service('passive_sonar/' + s[0], s[1], getattr(self, s[0])) for s in services.items()]
+
+    def bag_data(self, signals, trans, rot):
+        if self.bagging:
+            if signals.size == None:
+                self.signals_bag = self.signals_bag.reshape(0, signals.shape[0],
+                                                            signals.shape[1])
+            self.signals_bag = np.stack((self.signals_bag, [signals]))
+            self.trans_bag = np.stack((self.trans_bag, [trans]))
+            self.rot_bag = np.stack((self.rot_bag, [rot]))
+
+
+    def get_dtoa(self, signals):
+        '''
+        Returns a list of difference in time of arrival measurements for a signal between
+        each of the non_reference hydrophones and the single reference hydrophone.
+
+        signals - (self.receiver_count x self.signal_size) numpy array. It is assumed that the
+            first row of this array is the reference signal.
+        
+        returns: list of <self.receiver_count> dtoa measurements in units of microseconds.
+        '''
+        sampling_T = 1.0 / self.sampling_frequency
+        upsamp_T = sampling_T / self.upsampling_factor
+        t_max = sampling_T * signals.shape[1]
+
+        t = np.arange(0, t_max, step=sampling_T)
+        t_upsamp = np.arange(0, t_max, step=upsamp_T)
+
+        signals_upsamp = np.array([np.interp(t_upsamp, t, x) for x in signals])
+
+        dtoa, cross_corr, t_corr = \
+            map(np.array, zip(*[get_time_delta(t_upsamp, non_ref, signals_upsamp[0]) for non_ref \
+                                in signals_upsamp[1 : self.receiver_count]]))
+
+        t_corr = t_corr[0]  # should all be the same
+        self.visualize_dsp(t_upsamp, signals_upsamp, t_corr, cross_corr)
+
+        print "dtoa: {}".format(np.array(dtoa)*1E6)
+        return dtoa
+
+    #Passive Sonar Services
+
+    @thread_lock(lock)
+    def get_pulse_heading(self, srv):
+        '''
+        Returns the heading towards an active pinger emmiting at <self.target_frequency>.
+        Heading will be a unit vector in hydrophone_array frame
+        '''
+        success, err_str = True, ''
+        signals, p0, R = None, None, None
+        try:
+            time = rospy.Time.now()
+
+            # Poll input source until it reports it is ready
+            self.input_source.input_request()
+            while not self.input_source.ready:
+                if rospy.Time.now() - time > rospy.Duration(self.input_timeout):
+                    raise IOError('Timed out waiting for input source')
+                else:
+                    rospy.sleep(0.05)
+            self.input_source.reset()
+
+            # Gather input from source
+            signals, p0, R = self.input_source.get_input()
+
+            # Carry out multilateration to get heading to pinger
+            heading = self.multilaterator.getPulseLocation(self.get_dtoa(signals))
```

Comment:
This could be all zeros and throw an error, but it wouldn't be obvious why the pulse location wasn't valid when that divide by 0 error get's thrown. It would probably be good to add error logs in the `Multilaterator` class so you know why the estimate isn't valid.